# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/nsHttpChannel.cpp
# Commit: 55c0d49b821c
# Full Hash: 55c0d49b821c3bbe25f4bda4c4e5cca9e725287c
# Author: Paul Bone <pbone@mozilla.com>
# Date: 2019-11-01 16:13:11
# Regressor Bug: 1588412
# File Overlap Count: 2
# Description:
#   Bug 1588412 - (part 3) Remove MayChangeProcess calls fron nsHttpChannel r=mayhemer
#   
#   Also add assertions that enforce that fission requires
#   DocumentChannel.
#   
# ==============================================================================

diff -r e0e3a3390f76 -r 55c0d49b821c netwerk/protocol/http/nsHttpChannel.cpp
--- a/netwerk/protocol/http/nsHttpChannel.cpp	Thu Oct 31 18:36:56 2019 +0000
+++ b/netwerk/protocol/http/nsHttpChannel.cpp	Thu Oct 31 18:13:50 2019 +0000
@@ -2581,33 +2581,58 @@
       return NS_OK;
     }
 
-    nsCOMPtr<nsIParentChannel> parentChannel;
-    NS_QueryNotificationCallbacks(this, parentChannel);
-
-    RefPtr<DocumentChannelParent> documentChannelParent =
-        do_QueryObject(parentChannel);
-
-    if (!documentChannelParent) {
-      // notify "http-on-may-change-process" observers
-      gHttpHandler->OnMayChangeProcess(this);
-
-      if (mRedirectContentProcessIdPromise) {
-        MOZ_ASSERT(!mOnStartRequestCalled);
-
-        PushRedirectAsyncFunc(&nsHttpChannel::ContinueProcessResponse2);
-        rv = StartCrossProcessRedirect();
-        if (NS_SUCCEEDED(rv)) {
-          return NS_OK;
-        }
-        PopRedirectAsyncFunc(&nsHttpChannel::ContinueProcessResponse2);
-      }
-    }
+    AssertNotDocumentChennel();
   }
 
   // No process switch needed, continue as normal.
   return ContinueProcessResponse2(rv);
 }
 
+void nsHttpChannel::AssertNotDocumentChennel() {
+  if (!mLoadInfo || !IsDocument()) {
+    return;
+  }
+
+  nsCOMPtr<nsIParentChannel> parentChannel;
+  NS_QueryNotificationCallbacks(this, parentChannel);
+  RefPtr<DocumentChannelParent> documentChannelParent =
+      do_QueryObject(parentChannel);
+  if (documentChannelParent) {
+    // The load is using document channel.
+    return;
+  }
+
+  RefPtr<HttpChannelParent> httpParent = do_QueryObject(parentChannel);
+  if (!httpParent) {
+    // The load was initiated in the parent and doesn't need document
+    // channel.
+    return;
+  }
+
+  nsContentPolicyType contentPolicy;
+  MOZ_ALWAYS_SUCCEEDS(mLoadInfo->GetExternalContentPolicyType(&contentPolicy));
+  RefPtr<BrowsingContext> bc;
+  if (contentPolicy == CSPService::TYPE_DOCUMENT) {
+    MOZ_ALWAYS_SUCCEEDS(mLoadInfo->GetBrowsingContext(getter_AddRefs(bc)));
+  } else {
+    MOZ_ALWAYS_SUCCEEDS(mLoadInfo->GetFrameBrowsingContext(getter_AddRefs(bc)));
+  }
+  if (!bc) {
+    return;
+  }
+
+  if (mLoadInfo->LoadingPrincipal() &&
+      mLoadInfo->LoadingPrincipal()->IsSystemPrincipal()) {
+    // Loads with the system principal can skip document channel
+    return;
+  }
+
+  // The load was supposed to use document channel but didn't.
+  MOZ_DIAGNOSTIC_ASSERT(
+      !StaticPrefs::browser_tabs_documentchannel(),
+      "DocumentChannel is enabled but this load was done without it");
+}
+
 nsresult nsHttpChannel::ContinueProcessResponse2(nsresult rv) {
   if (NS_FAILED(rv) && !mCanceled) {
     // The process switch failed, cancel this channel.
@@ -7744,22 +7769,7 @@
       return NS_OK;
     }
 
-    nsCOMPtr<nsIParentChannel> parentChannel;
-    NS_QueryNotificationCallbacks(this, parentChannel);
-    RefPtr<DocumentChannelParent> documentChannelParent =
-        do_QueryObject(parentChannel);
-    if (!documentChannelParent) {
-      gHttpHandler->OnMayChangeProcess(this);
-
-      if (mRedirectContentProcessIdPromise) {
-        PushRedirectAsyncFunc(&nsHttpChannel::ContinueOnStartRequest1);
-        rv = StartCrossProcessRedirect();
-        if (NS_SUCCEEDED(rv)) {
-          return NS_OK;
-        }
-        PopRedirectAsyncFunc(&nsHttpChannel::ContinueOnStartRequest1);
-      }
-    }
+    AssertNotDocumentChennel();
   }
 
   // No process change is needed, so continue on to ContinueOnStartRequest1.